GITLAB - issue on softsewrve clone

# Overview

We need to add some mechanism to avoid softserve to try to clone too fast, this is breaking half of installations attempts. 


```bash 
Prepare Temporary Repo        ... done! [4 in 14m52.377s]
panic: error cloning config repository from soft serve

goroutine 1 [running]:
log.Panicf({0x1c38c47?, 0xc0000d4000?}, {0x0?, 0x1a?, 0x1?})
	/usr/local/go/src/log/log.go:392 +0x67
github.com/kubefirst/kubefirst/internal/softserve.configureSoftServe()
	/home/developer/app/internal/softserve/softserve.go:85 +0x214
github.com/kubefirst/kubefirst/internal/softserve.ConfigureSoftServeAndPush(0x0)
	/home/developer/app/internal/softserve/softserve.go:56 +0xbd
github.com/kubefirst/kubefirst/cmd.glob..func11(0xc00023a000?, {0x2bea110, 0x0, 0x0})
	/home/developer/app/cmd/createGitlab.go:99 +0x56f
github.com/kubefirst/kubefirst/cmd.glob..func9(0x2b9c800, {0x2bea110?, 0x0, 0x0})
	/home/developer/app/cmd/create.go:49 +0x1d6
github.com/spf13/cobra.(*Command).execute(0x2b9c800, {0x2bea110, 0x0, 0x0})
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0x2b9ca80)
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
github.com/kubefirst/kubefirst/cmd.Execute()
	/home/developer/app/cmd/root.go:26 +0x25
main.main()
	/home/developer/app/main.go:65 +0x41d
```

As the 1st action on softserve seems to be, to clone a config repo, maybe we can add a retry logic on this step until the clone work as expected or we give up after N attempts. 


On this code: 
https://github.com/kubefirst/kubefirst/blob/main/internal/softserve/softserve.go#L80

Instead of panicing, to add a retry logic.

--- a/internal/softserve/softserve.go
+++ b/internal/softserve/softserve.go
@@ -89,6 +89,7 @@ func ConfigureSoftServeAndPush(dryRun bool) {
 
 		success = true
 		log.Println("SoftServe successfully configured")
+		break
 	}
 
 	if !success {

--- a/internal/softserve/softserve.go
+++ b/internal/softserve/softserve.go
@@ -2,8 +2,8 @@ package softserve
 
 import (
 	"fmt"
-	"io/ioutil"
 	"log"
+	"os"
 	"strings"
 	"time"
 
@@ -41,39 +41,75 @@ func CreateSoftServe(dryRun bool, kubeconfigPath string) {
 
 }
 
+// ConfigureSoftServeAndPush calls Configure SoftServer and push gitops repository to GitLab
 func ConfigureSoftServeAndPush(dryRun bool) {
+
+	if dryRun {
+		log.Printf("[#99] Dry-run mode, configureSoftserveAndPush skipped.")
+		return
+	}
+
 	config := configs.ReadConfig()
 
 	configureAndPushFlag := viper.GetBool("create.softserve.configure")
 
-	if !configureAndPushFlag {
-		log.Println("Executing configureSoftserveAndPush")
-		if dryRun {
-			log.Printf("[#99] Dry-run mode, configureSoftserveAndPush skipped.")
-			return
+	// soft serve is already configured, skipping
+	if configureAndPushFlag {
+		log.Println("Skipping: configureSoftserveAndPush")
+		return
+	}
+
+	log.Println("Executing configureSoftserveAndPush")
+
+	success := false
+	totalAttempts := 5
+	for i := 0; i < totalAttempts; i++ {
+
+		log.Printf("Configuring SoftServe, attempt (%d of %d)", i+1, totalAttempts)
+
+		err := configureSoftServe()
+		if err != nil {
+			log.Printf("something went wrong, error is: %v, going to try again...", err)
+			time.Sleep(10 * time.Second)
+			continue
 		}
 
-		configureSoftServe()
-		// refactor: update it
 		gitlab.PushGitRepo(dryRun, config, "soft", "gitops")
 
 		viper.Set("create.softserve.configure", true)
-		viper.WriteConfig()
+		err = viper.WriteConfig()
+		if err != nil {
+			log.Printf("something went wrong, error is: %v, going to try again...", err)
+			time.Sleep(10 * time.Second)
+			continue
+		}
+
+		log.Println("waiting SoftServe to finish configuration, sleeping...")
 		time.Sleep(30 * time.Second)
-	} else {
-		log.Println("Skipping: configureSoftserveAndPush")
+
+		success = true
+		log.Println("SoftServe successfully configured")
 	}
+
+	if !success {
+		log.Panic("we tried hard to setup SoftServe but there were something wrong, please check the logs")
+	}
+
 }
 
-func configureSoftServe() {
+// configureSoftServe clones local repositories, update config.yaml local file and commit push to Soft Serve
+func configureSoftServe() error {
 	config := configs.ReadConfig()
 
-	url := "ssh://127.0.0.1:8022/config"
+	url := pkg.SoftServerURI
 	directory := fmt.Sprintf("%s/config", config.K1FolderPath)
 
 	log.Println("gitClient clone", url, directory)
 
-	auth, _ := pkg.PublicKey()
+	auth, err := pkg.PublicKey()
+	if err != nil {
+		return err
+	}
 
 	auth.HostKeyCallback = ssh2.InsecureIgnoreHostKey()
 
@@ -82,41 +118,52 @@ func configureSoftServe() {
 		Auth: auth,
 	})
 	if err != nil {
-		log.Panicf("error cloning config repository from soft serve")
+		return fmt.Errorf("error cloning config repository from soft serve, error: %v", err)
 	}
 
-	file, err := ioutil.ReadFile(fmt.Sprintf("%s/config.yaml", directory))
+	file, err := os.ReadFile(fmt.Sprintf("%s/config.yaml", directory))
 	if err != nil {
-		log.Panicf("error reading config.yaml file %s", err)
+		return fmt.Errorf("error reading config.yaml file %s", err)
 	}
 
 	newFile := strings.Replace(string(file), "allow-keyless: false", "allow-keyless: true", -1)
 
-	err = ioutil.WriteFile(fmt.Sprintf("%s/config.yaml", directory), []byte(newFile), 0)
+	err = os.WriteFile(fmt.Sprintf("%s/config.yaml", directory), []byte(newFile), 0)
 	if err != nil {
-		log.Panic(err)
+		return err
 	}
 
 	log.Printf("re-wrote config.yaml at %s/config folder", config.K1FolderPath)
 
-	w, _ := repo.Worktree()
+	w, err := repo.Worktree()
+	if err != nil {
+		return err
+	}
 
 	log.Println("Committing new changes...")
-	w.Add(".")
-	w.Commit("updating soft-serve server config", &git.CommitOptions{
+	_, err = w.Add(".")
+	if err != nil {
+		return err
+	}
+	_, err = w.Commit("updating soft-serve server config", &git.CommitOptions{
 		Author: &object.Signature{
 			Name:  "kubefirst-bot",
 			Email: config.InstallerEmail,
 			When:  time.Now(),
 		},
 	})
+	if err != nil {
+		return err
+	}
 
 	err = repo.Push(&git.PushOptions{
 		RemoteName: "origin",
 		Auth:       auth,
 	})
 	if err != nil {
-		log.Panic("error pushing to remote", err)
+		return fmt.Errorf("error pushing to remote, error: %v", err)
 	}
+
+	return nil
 
 }

--- a/pkg/constants.go
+++ b/pkg/constants.go
@@ -3,4 +3,5 @@ package pkg
 const (
 	ArgoCDLocalBaseURL = "https://localhost:8080/api/v1"
 	JSONContentType    = "application/json"
+	SoftServerURI      = "ssh://127.0.0.1:8022/config"
 )

